### PR TITLE
Egress Network policy fixes

### PR DESCRIPTION
**Pull request by pravisankar**
- We allow multiple policies to be created per namespace but we only recommend
  one policy per namespace as per our design.
  In case of multiple policies per namespace, we drop all corresponding ovs rules
  because its hard to determine which policy to apply first but currently we keep
  the latest policy firewall rules in this case. To be consistent with ovs rules
  behavior, this fix will drop all firewall rules as well.
- Renamed proxy.updateNetworkPolicy() to proxy.updateEgressNetworkPolicy()
  to make it more clear and not to get confused with np.updateNetworkPolicy()
  in networkpolicy.go

### CODE REVIEW

**Comment by pravisankar** [comment]
We also need to drop all egress firewall rules for global namespaces. This requires maintaining namespace to vnid map in the proxy. I will submit this change in a separate pr.

@openshift/networking PTAL

File: pkg/sdn/plugin/controller.go [hunk 1]
```diff
@@ -384,10 +384,12 @@ func (plugin *OsdnNode) updateEgressNetworkPolicyRules(vnid uint32) {
 	} else if vnid == 0 {
 		glog.Errorf("EgressNetworkPolicy in global network namespace is not allowed (%s); ignoring", policyNames(policies))
 	} else if len(namespaces) > 1 {
+		// Rationale: If multiple namespaces have their own egress network policy, it's hard to determine the order of policies to apply. Also it may be wrong to force a network policy on a namespace which only wants to join the network.
```
Comment (danwinship) [inline pkg/sdn/plugin/controller.go]: Mention something about VNIDs or NetNamespaces there since that's the source of the trouble; EgressNetworkPolicies are defined per-Namespace, but are implemented per-NetNamespace.

File: pkg/sdn/plugin/proxy.go [hunk 2]
```diff
@@ -87,29 +93,42 @@ func (proxy *OsdnProxy) watchEgressNetworkPolicies() {
...
 
-func (proxy *OsdnProxy) updateNetworkPolicy(policy osapi.EgressNetworkPolicy) {
-	firewall := make([]proxyFirewallItem, len(policy.Spec.Egress))
+func (proxy *OsdnProxy) updateEgressNetworkPolicy(policy osapi.EgressNetworkPolicy) {
+	if policy.Spec.Egress != nil {
+		if ref, ok := proxy.firewall[policy.Namespace]; ok {
+			if policy.UID != ref.policyUID {
+				// We only allow one policy per namespace otherwise it's hard to determine which policy to apply first
+				glog.Errorf("found multiple egress policies, dropping all firewall rules for namespace: %s", policy.Namespace)
+				delete(proxy.firewall, policy.Namespace)
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: Hm... this isn't right, since it means that the next time an ENP in that namespace gets updated, `proxy.firewall[policy.Namespace]` will be unset, and so that ENP *will* get applied (even if there are also still other ENPs in the namespace too).

We need to remember all the ENPs we've seen, but only apply the policy when there's just one in the Namespace.

**Comment by pravisankar** [comment]
@openshift/networking @danwinship updated, ptal

**Comment by pravisankar** [comment]
Added one more fix (new commit): SDN egress policy should not firewall endpoints from global namespaces.

**Comment by pravisankar** [comment]
[test]

File: pkg/sdn/plugin/proxy.go [hunk 3]
```diff
@@ -87,29 +98,68 @@ func (proxy *OsdnProxy) watchEgressNetworkPolicies() {
...
 			// should have been caught by validation
-			glog.Errorf("Illegal CIDR value %q in EgressNetworkPolicy rule", rule.To.CIDRSelector)
+			glog.Errorf("Illegal CIDR value %q in EgressNetworkPolicy rule for policy: %v", rule.To.CIDRSelector, policy.UID)
 			return
 		}
-		firewall[i] = proxyFirewallItem{rule.Type, cidr}
+		firewall[i] = firewallItem{rule.Type, cidr}
 	}
 
+	// Add/Update/Delete firwall rules for the namespace
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: typo "firwall"

File: pkg/sdn/plugin/proxy.go [hunk 4]
```diff
@@ -14,12 +14,23 @@ import (
...
+
+type policyFirewallItem struct {
+	uid      ktypes.UID
+	firewall []firewallItem
+}
+
 type proxyFirewallItem struct {
-	policy osapi.EgressNetworkPolicyRuleType
-	net    *net.IPNet
+	namespaceFirewalls []policyFirewallItem
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: it seems like the code below would be a lot simpler if this was a `map[ktypes.UID]policyFirewallItem` rather than a `[]policyFirewallItem`. (You might even be able to get rid of `policyFirewallItem` and just do `map[ktypes.UID][]firewallItem`?)

Also, maybe type/field naming would help make the code clearer. It's difficult to keep track of all the different "firewalls" and "items"...

File: pkg/sdn/plugin/proxy.go [hunk 5]
```diff
@@ -98,7 +103,39 @@ func (proxy *OsdnProxy) watchEgressNetworkPolicies() {
 	})
 }
 
+func (proxy *OsdnProxy) watchNetNamespaces() {
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: blah... we need to abstract out the EgressNetworkPolicy code better so that node and proxy can share more code rather than duplicating. Can you add a TODO comment for that?

File: pkg/sdn/plugin/proxy.go [hunk 6]
```diff
@@ -98,7 +103,39 @@ func (proxy *OsdnProxy) watchEgressNetworkPolicies() {
...
+		}
+		return nil
+	})
+}
 func (proxy *OsdnProxy) updateEgressNetworkPolicy(policy osapi.EgressNetworkPolicy) {
+	ns := policy.Namespace
+	ignorePolicy := false
+	func() {
+		proxy.idLock.Lock()
+		defer proxy.idLock.Unlock()
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: I would have a slight preference for splitting this out into a real helper function rather than an inline func() but if you really prefer it this way you can keep it

File: pkg/sdn/plugin/proxy.go [hunk 7]
```diff
@@ -111,7 +148,6 @@ func (proxy *OsdnProxy) updateEgressNetworkPolicy(policy osapi.EgressNetworkPoli
 	}
 
 	// Add/Update/Delete firwall rules for the namespace
-	ns := policy.Namespace
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: doesn't belong in this patch?

File: pkg/sdn/plugin/proxy.go (hunk 4 above)
Comment (dcbw) [inline pkg/sdn/plugin/proxy.go]: doing the map looks plausible and would make the code more readable, so +1 from me if it works

File: pkg/sdn/plugin/proxy.go (hunk 6 above)
Comment (dcbw) [inline pkg/sdn/plugin/proxy.go]: Could even do:

```
if func() bool {
    Lock()
    defer Unlock()
    if global vnid {
        log error
        return true
    }
    return false
}() {
    return
}
```

but a helper function is even clearer.

File: pkg/sdn/plugin/proxy.go [hunk 8]
```diff
@@ -87,29 +93,73 @@ func (proxy *OsdnProxy) watchEgressNetworkPolicies() {
...
+		firewall[i] = firewallItem{rule.Type, cidr}
 	}
 
+	// Add/Update/Delete firewall rules for the namespace
 	if len(firewall) > 0 {
-		proxy.firewall[policy.Namespace] = firewall
+		updated := false
+		if ref, ok := proxy.firewall[ns]; ok {
+			for uid := range ref.namespaceFirewalls {
+				if uid == policy.UID {
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: don't need a loop now that you're using a map; just `if _, ok := ref.namespaceFirewalls[policy.UID]; ok {`

Although you can get rid of `updated` in that case too. Just check if `proxy.firewall[ns]` exists, and create it if not, and then set `proxy.firewall[ns].namespaceFirewalls[policy.UID]`

File: pkg/sdn/plugin/proxy.go [hunk 9]
```diff
@@ -87,29 +93,73 @@ func (proxy *OsdnProxy) watchEgressNetworkPolicies() {
...
+			delete(proxy.firewall[ns].namespaceFirewalls, policy.UID)
+		}
+	}
+
+	// Set active policy for the namespace
+	if ref, ok := proxy.firewall[ns]; ok {
+		if len(ref.namespaceFirewalls) == 1 {
+			for uid := range ref.namespaceFirewalls {
+				ref.activePolicy = &uid
+				proxy.firewall[ns] = ref
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: If you made `proxy.firewall` be `map[string]*proxyFirewallItem` rather than `map[string]proxyFirewallItem` then you wouldn't need the reassignment here (and below).

File: pkg/sdn/plugin/proxy.go [hunk 10]
```diff
@@ -87,29 +93,73 @@ func (proxy *OsdnProxy) watchEgressNetworkPolicies() {
...
+				glog.Infof("Applied firewall egress network policy: %q to namespace: %q", uid, ns)
+			}
+		} else {
+			// We only allow one policy per namespace otherwise it's hard to determine which policy to apply first
+			glog.Errorf("Found multiple egress policies, dropping all firewall rules for namespace: %q", ns)
+			ref.activePolicy = nil
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: Hm... this is the opposite of the behavior in controller.go, where having multiple EgressNetworkPolicies means all traffic gets dropped, rather than all traffic being allowed.

**Comment by pravisankar** [comment]
@danwinship Updated, ptal

**Comment by danwinship** [comment]
[merge]
